econml/dml/dml.py

```diff
@@ -29,74 +29,85 @@
 from ..sklearn_extensions.model_selection import WeightedStratifiedKFold
 from ..utilities import (_deprecate_positional, add_intercept,
                          broadcast_unit_treatments, check_high_dimensional,
-                         cross_product, deprecated, fit_with_groups,
+                         cross_product, deprecated,
                          hstack, inverse_onehot, ndim, reshape,
                          reshape_treatmentwise_effects, shape, transpose,
                          get_feature_names_or_default, filter_none_kwargs)
 from .._shap import _shap_explain_model_cate
+from ..sklearn_extensions.model_selection import get_selector, ModelSelector, SingleModelSelector
 
 
-class _FirstStageWrapper:
-    def __init__(self, model, is_Y, featurizer, linear_first_stages, discrete_treatment):
-        self._model = clone(model, safe=False)
-        self._featurizer = clone(featurizer, safe=False)
-        self._is_Y = is_Y
-        self._linear_first_stages = linear_first_stages
-        self._discrete_treatment = discrete_treatment
-
-    def _combine(self, X, W, n_samples, fitting=True):
-        if X is None:
-            # if both X and W are None, just return a column of ones
-            return (W if W is not None else np.ones((n_samples, 1)))
-        XW = hstack([X, W]) if W is not None else X
-        if self._is_Y and self._linear_first_stages:
-            if self._featurizer is None:
-                F = X
-            else:
-                F = self._featurizer.fit_transform(X) if fitting else self._featurizer.transform(X)
-            return cross_product(XW, hstack([np.ones((shape(XW)[0], 1)), F]))
-        else:
-            return XW
+def _combine(X, W, n_samples):
+    if X is None:
+        # if both X and W are None, just return a column of ones
+        return (W if W is not None else np.ones((n_samples, 1)))
+    return hstack([X, W]) if W is not None else X
 
-    def fit(self, X, W, Target, sample_weight=None, groups=None):
-        if (not self._is_Y) and self._discrete_treatment:
-            # In this case, the Target is the one-hot-encoding of the treatment variable
-            # We need to go back to the label representation of the one-hot so as to call
-            # the classifier.
-            if np.any(np.all(Target == 0, axis=0)) or (not np.any(np.all(Target == 0, axis=1))):
-                raise AttributeError("Provided crossfit folds contain training splits that " +
-                                     "don't contain all treatments")
-            Target = inverse_onehot(Target)
 
-        if sample_weight is not None:
-            fit_with_groups(self._model, self._combine(X, W, Target.shape[0]), Target, groups=groups,
-                            sample_weight=sample_weight)
-        else:
-            fit_with_groups(self._model, self._combine(X, W, Target.shape[0]), Target, groups=groups)
-        return self
+class _FirstStageWrapper:
+    def __init__(self, model, discrete_target):
```

Comment:
I like the refactoring. There is a consequence of not being able to make model-specific (Y vs T) warnings though (like the binary outcomes PR tries to do in this class).

Reply:
Let's think about the cleanest way to do that as part of your PR; one thing to keep in mind is that we also use this wrapper for some Z models, so it's not just T vs. Y, which is what we assumed when we first implemented this logic when we only had DML but not the IV classes.